Lombiq.Tests.UI/Services/AtataFactory.cs

```diff
@@ -24,62 +25,79 @@ public static class AtataFactory
             .UseBaseUrl(baseUri.ToString())
             .UseCulture(browserConfiguration.AcceptLanguage.ToString())
             .UseTestName(configuration.AtataConfiguration.TestName)
-            .AddDebugLogging()
-            .AddLogConsumer(new TestOutputLogConsumer(testOutputHelper))
             .UseBaseRetryTimeout(timeoutConfiguration.RetryTimeout)
             .UseBaseRetryInterval(timeoutConfiguration.RetryInterval)
             .UseUtcTimeZone();
 
+        builder.LogConsumers.AddDebug();
+        builder.LogConsumers.Add(new TestOutputLogConsumer(testOutputHelper));
+
         configuration.AtataConfiguration.ContextBuilder?.Invoke(builder);
 
         return new AtataScope(builder.Build(), baseUri);
     }
+
+    public static void SetupShellCliCommandFactory() =>
+        ProgramCli.DefaultShellCliCommandFactory = OSDependentShellCliCommandFactory
+            .UseCmdForWindows()
+            .UseForOtherOS(new BashShellCliCommandFactory("-login"));
+
+    private static IWebDriver CreateDriver(
+        BrowserConfiguration browserConfiguration,
+        TimeoutConfiguration timeoutConfiguration,
+        ITestOutputHelper testOutputHelper)
+    {
+        IWebDriver From<T>(Func<BrowserConfiguration, TimeSpan, T> factory)
+            where T : IWebDriver =>
+            factory(browserConfiguration, timeoutConfiguration.PageLoadTimeout);
+
+        // Driver creation can fail with "Cannot start the driver service on http://localhost:56686/" exceptions if the
+        // machine is under load. Retrying it here so not the whole test needs to be re-run.
+        const int maxTryCount = 3;
+        var currentTry = 1;
+
+        // Force headless mode if we are in Linux without a working graphical environment.
+        if (!browserConfiguration.Headless && RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
+        {
+            browserConfiguration.Headless = string.IsNullOrWhiteSpace(Environment.GetEnvironmentVariable("DISPLAY"));
+        }
+
+        while (true)
+        {
+            try
+            {
+                return browserConfiguration.Browser switch
+                {
+                    Browser.Chrome => From(WebDriverFactory.CreateChromeDriver),
+                    Browser.Edge => From(WebDriverFactory.CreateEdgeDriver),
+                    Browser.Firefox => From(WebDriverFactory.CreateFirefoxDriver),
+                    Browser.InternetExplorer => From(WebDriverFactory.CreateInternetExplorerDriver),
+                    _ => throw new InvalidOperationException($"Unknown browser: {browserConfiguration.Browser}."),
+                };
+            }
+            catch (WebDriverException ex)
+            {
+                if (!ex.Message.ContainsOrdinalIgnoreCase("Cannot start the driver service on") || currentTry >= maxTryCount)
+                {
+                    throw;
+                }
+
+                currentTry++;
+                var retryCount = maxTryCount - currentTry + 1;
```

Comment:
On the first try, we should have 2 more tries, which would be:
```suggestion
                var retryCount = maxTryCount - currentTry;
```
Just like before.